edgehog-device-runtime-containers/src/docker/container.rs

```diff
@@ -0,0 +1,888 @@
+// This file is part of Edgehog.
+//
+// Copyright 2023 SECO Mind Srl
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+//! Docker struct to manage containers.
+
+use std::{
+    collections::HashMap,
+    fmt::{Debug, Display},
+    hash::Hash,
+    ops::{Deref, DerefMut},
+};
+
+use bollard::{
+    container::{
+        Config, CreateContainerOptions, InspectContainerOptions, NetworkingConfig,
+        RemoveContainerOptions, StartContainerOptions,
+    },
+    errors::Error as BollardError,
+    models::{
+        ContainerInspectResponse, EndpointSettings, HostConfig, PortBinding, RestartPolicy,
+        RestartPolicyNameEnum,
+    },
+};
+use tracing::{debug, info, instrument, trace, warn};
+
+use crate::{
+    client::*,
+    requests::{container::parse_port_binding, BindingError},
+};
+
+/// Error for the container operations.
+#[non_exhaustive]
+#[derive(Debug, thiserror::Error, displaydoc::Display)]
+pub enum ContainerError {
+    /// couldn't create the container
+    Create(#[source] BollardError),
+    /// couldn't inspect the container
+    Inspect(#[source] BollardError),
+    /// couldn't remove the container
+    Remove(#[source] BollardError),
+    /// couldn't start container.
+    Start(#[source] BollardError),
+    /// couldn't stop container
+    Stop(#[source] BollardError),
+    /// missing image reference in container definition
+    Image,
+}
+
+/// Docker container struct.
+#[derive(Debug, Clone, Eq)]
+pub struct Container<S>
+where
+    S: Hash + Eq,
+{
+    /// Id of the docker container.
+    pub id: Option<String>,
+    /// Assign the specified name to the container.
+    ///
+    /// Must match /?[a-zA-Z0-9][a-zA-Z0-9_.-]+.
+    pub name: S,
+    /// The name (or reference) of the image to use.
+    ///
+    /// The id of the image is optional since it will be available only when the image is.
```

Comment:
Is this comment correct? Because it seems it's indicating that image should be Option<S>